README
puppet module for managing tor
==============================

This module tries to manage tor, making sure it is installed, running, has munin
graphs if desired and allows for configuration of relays, hidden services, exit
policies, etc.

! Upgrade Notice !

 previously, if you did not set the $outbound_bindaddress variable, it was being 
 automatically set to the $listen_address variable. Now this is not being done
 and instead you will need to set the $outbound_bindaddress explicitly for it to 
 be set. 

 the tor::relay{} variables $bandwidth_rate and $bandwidth_burst were previously
 used for the tor configuration variables RelayBandwidthRate and
 RelayBandwidthBurst, these have been renamed to $relay_bandwidth_rate and
 $relay_bandwidth_burst. If you were using these, please rename your variables in
 your configuration.

 The variables $bandwidth_rate and $bandwidth_burst are now used for the tor
 configuration variables BandwidthRate and BandwidthBurst. If you used
 $bandwidth_rate or $bandwidth_burst please be aware that these values have
 changed and adjust your configuration as necessary.

 The $tor_ensure_version was converted to a parameter for the tor and
 tor::daemon classes.

 The $torsocks_ensure_version was converted to a parameter for the
 tor::torsocks class.

 The options that used to be settable with the
 tor::daemon::global_opts define now are parameters for the
 tor::daemon class, and tor::daemon::global_opts was
 removed accordingly.


Dependencies
============

This module needs:

- the concat module: git://labs.riseup.net/shared-concat

Usage
=====

Installing tor
--------------

To install tor, simply include the 'tor' class in your manifests:

    class { 'tor': }

You can specify the $ensure_version class parameter to get a specific
version installed.

However, if you want to make configuration changes to your tor daemon, you will
want to instead include the 'tor::daemon' class in your manifests, which will
inherit the 'tor' class from above:

    class { '::tor::daemon': }

You have the following class parameters that you can specify:

data_dir    (default: '/var/lib/tor')
config_file (default: '/etc/tor/torrc')
use_bridges (default: 0)
automap_hosts_on_resolve (default: 0)
log_rules   (default: ['notice file /var/log/tor/notices.log'])

The data_dir will be used for the tor user's $HOME, and the tor DataDirectory
value.

The config_file will be managed and the daemon restarted when
it changed.

use_bridges and automap_hosts_on_resolve are used to set the
UseBridges and AutomapHostsOnResolve torrc settings.

The log_rules can be an array of different Log lines, each will be added to the
config, for example the following will use syslog:

    class { '::tor::daemon':
        log_rules => [ 'notice syslog' ],
    }

If you want to set specific options for the tor class,
you may pass them directly to the tor::daemon in your manifests,
e.g.:

    class { '::tor::daemon':
      use_munin                 => true,
      automap_hosts_on_resolve  => 1,
    }

Configuring socks
-----------------

To configure tor socks support, you can do the following:

  tor::daemon::socks { "listen_locally": listen_addresses => [ '127.0.0.1' ]; }

this will setup the SocksListenAddress to be 127.0.0.1. You also can pass the
following options to tor::daemon::socks:

$port = 0       - SocksPort
$listen_address - can pass multiple values to configure SocksListenAddress lines
$policies       - can pass multiple values to configure SocksPolicy lines

Installing torsocks
-------------------

To install torsocks, simply include the 'torsocks' class in your manifests:

    class { 'torsocks': }

You can specify the $ensure_version class parameter to get a specific
version installed.

Configuring relays
==================

An example relay configuration:

 tor::daemon::relay { "foobar":
   port => 9001, listen_addresses => '192.168.0.1', address => '192.168.0.1',
   bandwidth_rate => '256', bandwidth_burst => '256', contact_info => "Foo <collective at example dot com>",
   my_family => '<long family string here>'
 }

You have the following options that can be passed to a relay, with the defaults shown:
 
$port                    = 0,
$listen_addresses        = [],
$portforwarding          = 0,     # PortForwarding 0|1, set for opening ports at the router via UPnP.
                                  # Requires 'tor-fw-helper' binary present.
$bandwidth_rate          = '',    # KB/s, defaulting to using tor's default: 5120KB/s
$bandwidth_burst         = '',    # KB/s, defaulting to using tor's default: 10240KB/s
$relay_bandwidth_rate    = 0,     # KB/s, 0 for no limit.
$relay_bandwidth_burst   = 0,     # KB/s, 0 for no limit.
$accounting_max          = 0,     # GB, 0 for no limit.
$accounting_start        = [],
$contact_info            = '',
$my_family               = '', # TODO: autofill with other relays
$address                 = "tor.${domain}",
$bridge_relay            = 0,
$ensure                  = present
$nickname                = $name

Configuring the control 
-----------------------

To pass parameters to configure the ControlPort and the HashedControlPassword,
you would do something like this:

 tor::daemon::control { "foo-control": 
  port => '80', hashed_control_password => '<somehash>',
  ensure => present 
}

Note: you must pass a hashed password to the control port, if you are going to
use it.


Configuring hidden services
---------------------------

To configure a tor hidden service you can do something like the following:

 tor::daemon::hidden_service { "hidden_ssh": ports => 22 }

The HiddenServiceDir is set to the ${data_dir}/${name}.

Configuring directories
-----------------------

An example directory configuration:

 tor::daemon::directory { 'ssh_directory':
   port => 80, listen_address => '192.168.0.1', 
   port_front_page => '/etc/tor/tor.html'
 }
  
Configuring exit policies
--------------------------

To configure exit policies, you can do the following:
 
tor::daemon::exit_policy { "ssh_exit_policy":
        accept => "192.168.0.1:22",
        reject => "*:*";
      }
    }


Polipo
======

Polipo support can be enabled by doing:

    include tor::polipo

this will inherit the tor class by default, remove privoxy if its installed, and
install polipo, making sure it is running.
  

Munin
=====

If you are using munin, and have the puppet munin module installed, you can set
the use_munin parameter to true when defining the tor::daemon class to have
graphs setup for you.



CHANGES.md
Platform 0.10
------------------------------------------------

The main focus for Platform 0.10 was to update of all client-side daemons to
newest releases, like Soledad and OpenVPN. This introduces a *compatibility
change*: by setting the platform version to 0.10, it also requires client 0.9.4
or later. We also switched the development branch to the 'master' branch and are
creating a branch called 0.10.x to push hot-fixes during the 0.10 life-cycle.

Note: This will be the last major release of the LEAP Platform for Debian
Jessie. We will continue to support 0.10 with minor releases with important
security and bug fixes, but the next major release will require an upgrade to
Stretch.

New Features:

* Tor single-hop onion service capability.
* `leap info` is now run after deploy
* Timestamps are added to deployments
* Missing ssh host keys are generated on node init
* Private networking support for local Vagrant development
* Static sites get lets encrypt support
* add command `leap node disable`, `leap node enable`, `leap ping`

Notable Changes:

* Removed haproxy because we don't support multi-node couchdb installations anymore (#8144).
* Disable nagios notification emails (#8772).
* Fix layout of apt repository (#8888)
* Limit what archive signing keys are accepted for the leap debian repository packages (#8425).
* Monitor the Webapp logs for errors (#5174).
* Moved development to the master branch.
* Rewrite leap_cli ssh code
* Debian wheezy was fully deprecated
* Restructure package archives to enable auto packaging, and CI testing
* Significant CI improvements
* Troubleshooting information added to `leap user ls`
* Couchdb service is no longer required on soledad nodes (#8693)
* Tor service refactored (#8864), and v3 hidden service support added (#8879)
* Fixed unattended-upgrades (#8891)
* Alert on 409 responses for webapp
* Many other issues resolved, full list: https://0xacab.org/groups/leap/milestones/platform-010?title=Platform+0.10

Upgrading:

If you have a node with the service 'tor' defined, you will need to change it to
be either 'tor-relay', or 'tor-exit'. Look in your provider directory under the
nodes directory for any .json file that has a 'services' section with 'tor'
defined, change that to the correct tor service you are wanting to deploy.

Make sure you have the correct version of leap_cli

    workstation$ sudo gem install leap_cli --version=1.9

If you are upgrading from a version previous to 0.9, please follow those upgrade
instructions before upgrading to 0.10.

Prepare your platform source by checking out the 0.10.x branch:

    workstation$ cd leap_platform
    workstation$ git fetch
    workstation$ git checkout 0.10.x

Then, deploy:

    workstation$ cd $PROVIDER_DIR
    workstation$ leap deploy
    workstation$ leap test

After deployment, if the leap test does not succeed, you should
investigate. Please see below for some post-deployment upgrade steps that you
may need to perform.

Starting with Soledad Server 0.9.0, the CouchDB database schema was changed to
improve speed of the server side storage backend. If you provided email, you
will need to run the migration script, otherwise it is unnecessary. Until you
migrate, soledad will refuse to start.

To run the migration script, do the following (replacing $PROVIDER_DIR,
$COUCHDB_NODE, $MX_NODE, and $SOLEDAD_NODE with your values):

First backup your couchdb databases, just to be safe. NOTE: This can take some
time and will place several hundred megabytes of data into
/var/backups/couchdb. The size and time depends on how many users there are on
your system. For example, 15k users took approximately 25 minutes and 308M of
space:

    workstation$ leap ssh $COUCHDB_NODE
    server# cd /srv/leap/couchdb/scripts
    server# ./cleanup-user-dbs
    server# time ./couchdb_dumpall.sh

 Once that has finished, then its time to run the migration:

    workstation$ cd $PROVIDER_DIR
    workstation$ leap run 'systemctl leap_mx stop' $MX_NODE
    workstation$ leap run --stream '/usr/share/soledad-server/migration/0.9/migrate.py --log-file /var/log/leap/soledad_migration --verbose --do-migrate' $SOLEDAD_NODE
    wait for it to finish (will print DONE)
    rerun if interrupted
    workstation$ leap deploy
    workstation$ leap test

Known Issues:

If you have been deploying from our master branch (ie: unstable code), you might
end up with a broken sources line for apt. If you get the following:
    WARNING: The following packages cannot be authenticated!

Then you should remove the files on your nodes inside
/var/lib/puppet/modules/apt/keys and deploy again. (#8862, #8876)

* When upgrading, sometimes systemd does not report the correct state of a
  daemon. The daemon will be not running, but systemd thinks it is. The symptom
  of this is that a deploy will succeed but `leap test` will fail. To fix, you
  can run `systemctl stop DAEMON` and then `systemctl start DAEMON` on the
  affected host (systemctl restart seems to work less reliably).

Includes:

* leap_web: 0.9.2
* nickserver: 0.10.0
* leap-mx: 0.10.1
* soledad-server: 0.10.5

Commits: https://0xacab.org/groups/leap/milestones/platform-010?title=Platform+0.10

For details on about all the changes included in this release please consult the
[LEAP platform 0.10 milestone](https://0xacab.org/leap/platform/milestones/7 ).


Platform 0.9
--------------------------------------

The focus for Platform 0.9 was to clean house: we replaced the annoying system
of puppet submodules, we cleaned up the directory structure, we removed many of
the gem dependencies, and we fixed a lot of bugs.

New Features:

* `leap vm` -- Support for managing remote virtual servers (AWS only, for now)
* `leap cert renew` -- Integration with Let's Encrypt
* `leap open monitor` -- for handy access to nagios
* improved documentation -- open docs/index.html to see

Notable Changes:

* 86 bugs fixed
* Fixed security issues with VPN
* More tests
* Replaced git submodules with git subrepo

platform.rb
# encoding: utf-8
#
# These are variables defined by this leap_platform and used by leap_cli.
#

Leap::Platform.define do
  self.version = "0.10"
  self.compatible_cli = "1.9".."1.99"

  #
  # the facter facts that should be gathered
  #
  self.facts = ["ec2_local_ipv4", "ec2_public_ipv4"]

  #
  # absolute paths on the destination server
  #
  self.hiera_dir  = '/etc/leap' if self.respond_to?(:hiera_dir)
  self.hiera_path = '/etc/leap/hiera.yaml'
  self.leap_dir   = '/srv/leap'
  self.files_dir  = '/srv/leap/files'
  self.init_path  = '/srv/leap/initialized'

  #
  # the named paths for this platform
  # (relative to the provider directory)
  #
  self.paths = {
    # directories
    :hiera_dir        => 'hiera',
    :files_dir        => 'files',
    :nodes_dir        => 'nodes',
    :services_dir     => 'services',
    :templates_dir    => 'templates',
    :tags_dir         => 'tags',
    :node_files_dir   => 'files/nodes/#{arg}',

    # input config files
    :common_config    => 'common.json',
    :provider_config  => 'provider.json',
    :service_config   => 'services/#{arg}.json',
    :tag_config       => 'tags/#{arg}.json',
    :template_config  => 'templates/#{arg}.json',
    :secrets_config   => 'secrets.json',
    :cloud_config     => 'cloud.json',
    :node_config      => 'nodes/#{arg}.json',

    # input config files, environmentally scoped
    :common_env_config    => 'common.#{arg}.json',
    :provider_env_config  => 'provider.#{arg}.json',
    :service_env_config   => 'services/#{arg[0]}.#{arg[1]}.json',
    :tag_env_config       => 'tags/#{arg[0]}.#{arg[1]}.json',

    # input templates
    :provider_json_template        => 'files/service-definitions/provider.json.erb',
    :eip_service_json_template     => 'files/service-definitions/#{arg}/eip-service.json.erb',
    :soledad_service_json_template => 'files/service-definitions/#{arg}/soledad-service.json.erb',
    :smtp_service_json_template    => 'files/service-definitions/#{arg}/smtp-service.json.erb',

    # custom files
    :custom_puppet_dir => 'files/puppet',
    :custom_puppet_modules_dir => 'files/puppet/modules',
    :custom_puppet_manifests_dir => 'files/puppet/manifests',
    :custom_tests => 'files/tests',
    :custom_bin => 'files/bin',

    # output files
    :facts            => 'facts.json',
    :user_dir         => 'users/#{arg}',
    :user_ssh         => 'users/#{arg}/#{arg}_ssh.pub',
    :user_pgp         => 'users/#{arg}/#{arg}_pgp.pub',
    :known_hosts      => 'files/ssh/known_hosts',
    :authorized_keys  => 'files/ssh/authorized_keys',
    :monitor_pub_key  => 'files/ssh/monitor_ssh.pub',
    :monitor_priv_key => 'files/ssh/monitor_ssh',
    :ca_key           => 'files/ca/ca.key',
    :ca_cert          => 'files/ca/ca.crt',
    :client_ca_key    => 'files/ca/client_ca.key',
    :client_ca_cert   => 'files/ca/client_ca.crt',
    :dh_params        => 'files/ca/dh.pem',
    :acme_key         => 'files/ca/lets-encrypt-account.key',
    :acme_info        => 'files/ca/lets-encrypt-account.json',
    :commercial_key   => 'files/cert/#{arg}.key',
    :commercial_csr   => 'files/cert/#{arg}.csr',
    :commercial_cert  => 'files/cert/#{arg}.crt',
    :dkim_priv_key    => 'files/mx/dkim.key',
    :dkim_pub_key     => 'files/mx/dkim.pub',

    :commercial_ca_cert       => 'files/cert/commercial_ca.crt',
    :vagrantfile              => 'test/Vagrantfile',
    :static_web_provider_json => 'files/web/bootstrap/#{arg}/provider.json',
    :static_web_htaccess      => 'files/web/bootstrap/#{arg}/htaccess',
    :static_web_readme        => 'files/web/bootstrap/README',

    # node output files
    :hiera             => 'hiera/#{arg}.yaml',
    :node_ssh_pub_key  => 'files/nodes/#{arg}/#{arg}_ssh.pub',
    :node_x509_key     => 'files/nodes/#{arg}/#{arg}.key',
    :node_x509_cert    => 'files/nodes/#{arg}/#{arg}.crt',
    :node_tor_priv_key => 'files/nodes/#{arg}/tor.key',
    :node_tor_pub_key  => 'files/nodes/#{arg}/tor.pub',

    # testing files
    :test_client_key     => 'test/cert/client.key',
    :test_client_cert    => 'test/cert/client.crt',
    :test_openvpn_config => 'test/openvpn/#{arg}.ovpn',
    :test_client_openvpn_template => 'test/openvpn/client.ovpn.erb'
  }

  #
  # the files that need to get renamed when a node is renamed
  #
  self.node_files = [
    :node_config, :hiera, :node_x509_cert, :node_x509_key, :node_ssh_pub_key
  ]

  self.monitor_username = 'monitor'

  self.reserved_usernames = ['monitor', 'root']

  self.default_puppet_tags = ['leap_base','leap_service']
end

